### PR TITLE
optimize integration testing in CI

### PR DESCRIPTION
**Pull request by nashif**
This adds a new option, integration which is used to list platforms to use in integration testing (CI) and avoids whitelisting platforms. This will enable running those tests on real hardware if the --integration option is omitted.
--integration option is now not default, in the future this will be the default and will build/test a minimal set, running sanitycheck with more coverage will be possible use another option.

oh, one more thing in this PR: use native_posix exclusively where it makes sense and avoid massive build/running trivial stuff that is being tested elsewhere anyways..

### CODE REVIEW

**Review by galak — COMMENTED** [review]
How about the YAML being `integration_board`?  The mix in sanitycheck of platform and our use of board everywhere else has always been a little funny.

**Comment by nashif** [comment]
stats:

current master:
```
i9:zephyr(master): sanitycheck -y
Renaming output directory to /home/nashif/Work/zephyrproject/zephyr/sanity-out.5
INFO    - JOBS: 20
INFO    - Selecting default platforms per test case
INFO    - Building initial testcase list...
INFO    - 4206 test configurations selected, 202630 configurations discarded due to filters.
INFO    - Completed in 3 seconds
```

this PR with no options:
```
i9:zephyr(sanitycheck/optimize_tests): sanitycheck -y
Renaming output directory to /home/nashif/Work/zephyrproject/zephyr/sanity-out.3
INFO    - JOBS: 20
INFO    - Selecting default platforms per test case
INFO    - Building initial testcase list...
INFO    - 4444 test configurations selected, 202392 configurations discarded due to filters.
INFO    - Completed in 3 seconds
```

this PR with -G
```
i9:zephyr(sanitycheck/optimize_tests): sanitycheck -y -G
Renaming output directory to /home/nashif/Work/zephyrproject/zephyr/sanity-out.4
INFO    - JOBS: 20
INFO    - Selecting default platforms per test case
INFO    - Building initial testcase list...
INFO    - 4033 test configurations selected, 202803 configurations discarded due to filters.
INFO    - Completed in 3 seconds
```

File: samples/sensor/lps22hb/sample.yaml [hunk 1]
```diff
@@ -4,7 +4,8 @@ tests:
   sample.sensor.lps22hb:
     harness: console
     tags: sensors
-    depends_on: i2c lps22hb
+    pre-filter: subsys_dt_compat_enabled("i2c")
```
Comment (erwango) [inline samples/sensor/lps22hb/sample.yaml:7]: I wonder what is the difference between this and the above change:
```
- depends_on: i2c lis2dh
+ pre-filter: subsys_dt_compat_enabled("i2c") and dt_compat_enabled("st,lis2dh")
```

**Comment by erwango** [comment]
@nashif do you think it would be feasible to override integration boards locally ? (I'm thinking about a new prop in map.yaml)

**Comment by nashif** [comment]
> @nashif do you think it would be feasible to override integration boards locally ? (I'm thinking about a new prop in map.yaml)

why would you want to do that? you can still run on any boards with full coverage by specifying the boards directly.

File: samples/sensor/lps22hb/sample.yaml (hunk 1 above)
Comment (nashif) [reply samples/sensor/lps22hb/sample.yaml:7]: the above change probably need to be changed, i2c in pre-filter, and lis2dh in filter.

**Comment by erwango** [comment]
> why would you want to do that? you can still run on any boards with full coverage by specifying the boards directly.

Maybe I'm not quite clear with the new option. And reading again the description, I'm actually more confused. I understand the interest of removing the withelisting, but I'm not clear why the integration makes things better.
Can you provide more details on the problematic you're tackling here ?

**Comment by nashif** [comment]
> Can you provide more details on the problematic you're tackling here ?

--integration helps optimize the coverage for CI runs, everything else is kept the same or is improved by relying on up2date information from dts re hardware availability and support.

The goal is not to randomly build and run tests and instead do this in the most optimal way balance coverage with run-time required in CI when someone submits a PR.

**Comment by erwango** [comment]
> The goal is not to randomly build and run tests and instead do this in the most optimal way balance coverage with run-time required in CI when someone submits a PR.

Get it better, txs

**Comment by nvlsianpu** [comment]
Did I understand this patch well: apart form what listed under integration, qemu_x86 will be always one of  CI target as well (if filters mach?).
Currently settings, storage and FS are tested in CI on either qemu and posix (which is great, and i really like as extensive testing are problematic for flash endurance). I wan't to avoid disabling these test in CI accidentally.

BTW: some of commits here are helpers patches, but not strongly connected and can go to the master independently, even right now.

**Review by tyrantlizard77 — COMMENTED** [review]
can you come up with something shorter than `subsys_dt_compat_enabled` for those of us who don't like to type? literally every `pre-filer:` statement uses it

**Comment by nashif** [comment]
superseded with #28435

**Pull request closed without merging**